sys/net/if_ethersubr.c

```diff
@@ -1439,6 +1439,7 @@ ether_8021q_frame(struct mbuf **mp, struct ifnet *ife, struct ifnet *p,
 	 * If PCP is set in mbuf, use it
 	 */
 	if ((*mp)->m_flags & M_VLANTAG) {
+		vid = EVL_VLANOFTAG((*mp)->m_pkthdr.ether_vtag);
```

Comment:
This is wrong. The PCP set in mbuf is from upper layer, aka from ip_output(). It will be layer violation for Upper layer to set vlan id in mbuf.

Reply:
Then IP layer should stop write PCP to Ethernet layer.

Reply:
> This is wrong. The PCP set in mbuf is from upper layer, aka from ip_output(). It will be layer violation for Upper layer to set vlan id in mbuf.

Regardless the layer violation, this will break vlan(4). The vlan(4) relies `ether_8021q_frame()` to properly inject VID.

Reply:
> Then IP layer should stop write PCP to Ethernet layer.
That was introduced via 868aabb4708d to support priority per-flow . 

> Upper layer will always write VID=0, as it hardcoded or respect already set VID.
I do not quite follow. I do not see any issues with that, to always write VID=0.

> or respect already set VID.
Upper layers should not do that. From the perspective from upper layers, the VID does not exist, or it can be treated to be 0, exactly.

Reply:
Current code base state violates ```m_pkthdr.ether_vtag``` usage.
Current implementation of ip_output() does respect ```m_pkthdr.ether_vtag``` by using ```EVL_APPLY_PRI()``` macro.

Only ```ether_8021q_frame()``` does respect nothing.

Reply:
Current ```ether_set_pcp()``` set PCP + zeroize VID and CFI.

Reply:
> Current code base state violates `m_pkthdr.ether_vtag` usage.

For inbound packets, `M_VLANTAG` and `ether_vtag` is used by drivers to indicate that the hardware / firmware / driver has processed the 802.1Q tag.

For outbound traffic,  `M_VLANTAG` is used to direct the driver / firmware / hardware to do 802.1Q tag inserting.

> Current implementation of ip_output() does respect `m_pkthdr.ether_vtag` by using `EVL_APPLY_PRI()` macro.

This should be the only violation in current net stack.

> 
> Only `ether_8021q_frame()` does respect nothing.

> Current ether_set_pcp() set PCP + zeroize VID and CFI.

That is on purpose, to support PCP for untagged ( NULL VLAN ID ) outbound traffic.

Reply:
> That is on purpose, to support PCP for untagged ( NULL VLAN ID ) outbound traffic.

This should not go in this way, like ip_output does not force VID=0, CFI=0.

Any way, I make simple implementation that respect ```ether_vtag``` changes from any code path, and now code more clear and predictable.
I need some time to check that it work.

Reply:
It work with if_vlan and code base from company product.